UM/Resources.py

```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
+        config_path = None
         if platform.system() == "Windows":
-            cls.__config_storage_path = os.path.join(os.getenv("APPDATA"), storage_dir_name)
-            # cache is machine-specific so we put it in "LOCAL"
-            cls.__cache_storage_path = os.path.join(os.getenv("LOCALAPPDATA"), storage_dir_name, "cache")
+            config_path = os.getenv("APPDATA")
         elif platform.system() == "Darwin":
-            cls.__config_storage_path = os.path.join(os.path.expanduser("~/Library/Application Support"), storage_dir_name)
-            # For backward compatibility, support loading files from the old storage location
-            cls.addSearchPath(os.path.expanduser("~/.{0}".format(storage_dir_name)))
+            config_path = os.path.expanduser("~/Library/Application Support")
         elif platform.system() == "Linux":
-            xdg_config_home = ""
             try:
-                xdg_config_home = os.environ["XDG_CONFIG_HOME"]
+                config_path = os.environ["XDG_CONFIG_HOME"]
             except KeyError:
-                xdg_config_home = os.path.expanduser("~/.config")
-            cls.__config_storage_path = os.path.join(xdg_config_home, storage_dir_name)
+                config_path = os.path.expanduser("~/.config")
+        else:
+            config_path = "."
 
-            xdg_data_home = ""
+        return config_path
+
+    @classmethod
+    def getPossibleConfigStorageRootPathList(cls):
+        # Returns all possible root paths for storing app configurations (in old and new versions)
+        config_root_list = [Resources.getConfigStorageRootPath()]
+        if platform.system() == "Windows":
+            # it used to be in LOCALAPPDATA on Windows
+            config_root_list.append(os.getenv("LOCALAPPDATA"))
+
+        config_root_list = [os.path.join(n, cls.ApplicationIdentifier) for n in config_root_list]
+        return config_root_list
+
+    @classmethod
+    def getDataStorageRootPath(cls):
+        # Returns the path where we store different versions of app data
+        data_path = None
+        if platform.system() == "Linux":
             try:
-                xdg_data_home = os.environ["XDG_DATA_HOME"]
+                data_path = os.environ["XDG_DATA_HOME"]
             except KeyError:
-                xdg_data_home = os.path.expanduser("~/.local/share")
-            cls.__data_storage_path = os.path.join(xdg_data_home, storage_dir_name)
+                data_path = os.path.expanduser("~/.local/share")
+        return data_path
 
-            xdg_cache_home = ""
+    @classmethod
+    def getCacheStorageRootPath(cls):
+        # Returns the path where we store different versions of app configurations
+        cache_path = None
+        if platform.system() == "Windows":
+            cache_path = os.getenv("LOCALAPPDATA")
+        elif platform.system() == "Darwin":
+            cache_path = None
```

Comment:
There are some fuzzy logic in the old code too. Basically, there must be a path for the **config dir**. If no place can be found for it, it will be `.` the current directory. But for **data** and **cache**, it depends. Only on Linux do we have separate places for all of them; On Windows **config** and **data** are kept in the Roaming, and **cache** in Local; On Mac all of them are in the same place. This basically means **data** and **cache** are optional to have their own places, thus the `None` in `getCache` and `getData`.